uitools/widgets/Esri/ArcGISRuntime/Toolkit/FloorFilter.cpp

```diff
@@ -0,0 +1,313 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "FloorFilter.h"
+
+#include "ui_FloorFilter.h"
+
+// Toolkit headers
+#include "FloorFilterController.h"
+#include "FloorFilterFacilityItem.h"
+#include "FloorFilterLevelItem.h"
+#include "FloorFilterSiteItem.h"
+
+// ArcGISRuntime headers
+#include <FloorFacility.h>
+#include <FloorSite.h>
+#include <MapGraphicsView.h>
+#include <SceneGraphicsView.h>
+
+// Qt headers
+#include <QEvent>
+#include <QSortFilterProxyModel>
+
+namespace Esri {
+namespace ArcGISRuntime {
+namespace Toolkit {
+
+  namespace {
+
+    /*!
+     \internal
+     \brief Given a \a modelIndex, for \a model,
+     extracts the `T*` pointer from its \c userRole.
+     */
+    template <typename T>
+    T* itemForIndex(QAbstractItemModel* model, QModelIndex modelIndex)
+    {
+      if (!model)
+        return nullptr;
+
+      if (modelIndex.isValid())
+      {
+        return qvariant_cast<T*>(model->data(modelIndex, Qt::UserRole));
+      }
+      return nullptr;
+    }
+
+    /*!
+     \internal
+     \brief Given some modelId in a given model, returns that items index in the model.
+            O(n) search time.
+     */
+    template <typename T>
+    QModelIndex indexForId(QAbstractItemModel* model, QString id)
+    {
+      const int rowCount = model->rowCount();
+      for (int i = 0; i < rowCount; ++i)
+      {
+        const auto index = model->index(i, 0);
+        const auto item = itemForIndex<T>(model, index);
+        if (item && item->modelId() == id)
+          return index;
+      }
+      return QModelIndex{};
+    }
+
+  /*!
+    \internal
+    \brief Wrap the \a sourceModel with a \l QSortFilterProxyModel that sorts and filters based on the \a filterField textEdited event.
+   */
+    QSortFilterProxyModel* wrapFilterModel(QAbstractItemModel* sourceModel, QLineEdit* filterField, FloorFilter* parent)
+    {
+      auto model = new QSortFilterProxyModel(parent);
+      model->setSourceModel(sourceModel);
+      model->setSortCaseSensitivity(Qt::CaseInsensitive);
+      model->setFilterCaseSensitivity(Qt::CaseInsensitive);
+
+      QObject::connect(filterField, &QLineEdit::textEdited, parent, [model](const QString& text)
+                       {
+                         model->setFilterFixedString(text);
+                       });
+      return model;
+    }
+
+    /*!
+     \internal
+     \brief Helper struct that holds a reference to sme `b`, sets `b` to `v` on construction,
```

Comment:
"to some `b`"?

Reply:
b and v are not very meaningful names to me. Maybe something like "original_value" and "completion_value" is a bit clearer